middleware/common/include/common/flag.h

```diff
@@ -6,176 +6,168 @@
 
 #pragma once
 
-#include "common/serialize/macro.h"
-#include "common/string.h"
-#include "common/traits.h"
+#include "common/algorithm.h"
+#include "common/stream/customization.h"
 
-#include "common/code/raise.h"
-#include "common/code/casual.h"
-
-#include <initializer_list>
-#include <type_traits>
+#include <bitset>
 
 namespace casual
 {
-   namespace common
+   namespace common::flag
    {
-
-      namespace has
+      template< typename T>
+      concept enum_flag_like = concepts::enumerator< T> && requires( T e)
       {
-         template< auto flags, typename T>
-         constexpr inline bool flag( T value)
-         {
-            return ( value & flags) == flags;
-         }
-      } // has
-
+         casual_enum_as_flag( e);
+      };
 
-      template< typename E>
-      struct Flags
+      namespace detail
       {
-         using enum_type = E;
-         using underlying_type = std::underlying_type_t< enum_type>;
-
-         static_assert( std::is_enum< enum_type>::value, "E has to be of enum type");
+         template< typename T>
+         concept has_description = enum_flag_like< T> && requires( T e)
+         {
+            { description( e)} -> std::same_as< std::string_view>;
+         };
          
+      } // detail
+      
+   } // common::flag
+   
+} // casual
 
-         //! just a helper to easier deduce the enum-type
-         constexpr static auto type() noexcept { return enum_type{};}         
-
-         constexpr Flags() = default;
-
-         template< typename... Enums>
-         constexpr Flags( enum_type e, Enums... enums) noexcept : Flags( bitmask( e, enums...)) {}
- 
-         constexpr explicit Flags( underlying_type flags) noexcept : m_flags( flags) {}
-
-         constexpr Flags convert( underlying_type flags) const
-         {
-            if( flags & ~underlying())
-               code::raise::error( code::casual::invalid_argument, "flags: ", flags, " limit: ", *this);
+template< casual::common::flag::enum_flag_like T>
+constexpr auto operator ! ( T value)
+{
+   return std::to_underlying( value) == 0;
+}
 
```

Comment:
Had to do some thinking about what this did/should do. My first thougt was incorrect, My first thought was that operator! for a flag like object should do som kind of flag inversion ("complement"), which is not reasonable.

I think this does the same thing as common>::flag::empty().

Reply:
Yeah,  this might not be a good operator to overload. The thinking was be able to do `if( ! flag)`, but since you can't do `if( flag)` it's not symmetrical. We might be better of not to implement `! operator`